packages/contracts-bedrock/src/L1/SystemConfig.sol

```diff
@@ -141,10 +213,85 @@ contract SystemConfig is OwnableUpgradeable, Semver {
+    /// @param _addr The address to store
+    /// @param _slot The storage slot to store the address in.
+    /// @dev WARNING! This function must be used cautiously, as it allows for overwriting values
+    ///      in arbitrary storage slots. Solc will add checks that the data passed as `_addr`
+    ///      is 20 bytes or less.
+    function _setAddress(address _addr, bytes32 _slot) internal {
+        bytes32 slot = _slot;
+        assembly {
+            sstore(slot, _addr)
+        }
+    }
+
+    /// @notice Returns an address stored in an arbitrary storage slot.
+    ///         These storage slots decouple the storage layout from
+    ///         solc's automation.
+    /// @param _slot The storage slot to retrieve the address from.
+    function _getAddress(bytes32 _slot) internal view returns (address) {
+        address addr;
+        bytes32 slot = _slot;
+        assembly {
+            addr := sload(slot)
+        }
+        return addr;
+    }
+
+    /// @notice Getter for the L1CrossDomainMessenger address.
+    function l1CrossDomainMessenger() external view returns (address) {
+        return _getAddress(L1_CROSS_DOMAIN_MESSENGER_SLOT);
+    }
+
+    /// @notice Getter for the L1ERC721Bridge address.
+    function l1ERC721Bridge() external view returns (address) {
+        return _getAddress(L1_ERC_721_BRIDGE_SLOT);
+    }
+
+    /// @notice Getter for the L1StandardBridge address.
+    function l1StandardBridge() external view returns (address) {
+        return _getAddress(L1_STANDARD_BRIDGE_SLOT);
+    }
+
+    /// @notice Getter for the L2OutputOracle address.
+    function l2OutputOracle() external view returns (address) {
+        return _getAddress(L2_OUTPUT_ORACLE_SLOT);
+    }
+
+    /// @notice Getter for the OptimismPortal address.
+    function optimismPortal() external view returns (address) {
+        return _getAddress(OPTIMISM_PORTAL_SLOT);
+    }
+
+    /// @notice Getter for the OptimismMintableERC20Factory address.
+    function optimismMintableERC20Factory() external view returns (address) {
+        return _getAddress(OPTIMISM_MINTABLE_ERC20_FACTORY_SLOT);
+    }
+
+    /// @notice Getter for the BatchInbox address.
+    function batchInbox() external view returns (address) {
+        return _getAddress(BATCH_INBOX_SLOT);
+    }
+
+    /// @notice Sets the start block in a backwards compatible way. Proxies
+    ///         that were initialized before this existed will not have it
+    ///         be set by `block.number`.
+    function _setStartBlock(uint256 _startBlock) internal {
+        if (_startBlock != 0) {
+            // There is an override, it cannot already be set.
+            require(startBlock == 0, "SystemConfig: cannot override an already set start block");
+            startBlock = _startBlock;
+        } else if (startBlock == 0) {
+            // There is no override and it is not set in storage. Set it to the block number.
+            startBlock = block.number;
+        }
+    }
```

Comment:
Can we change this into the following? The current dispatch logic is a little wonky to reason about. Note, the following changes one codepath whereby currently if the `_startBlock == 0` and `startBlock != 0`, nothing will happen, in the following suggestion, the top-level require statement will cause a revert if the `startBlock != 0` (even if `_startBlock == 0`)
```suggestion
    /// @notice Sets the start block in a backwards compatible way. Proxies
    ///         that were initialized before this existed will not have it
    ///         be set by `block.number`.
    function _setStartBlock(uint256 _startBlock) internal {
        require(startBlock == 0, "SystemConfig: cannot override an already set start block");
        if (_startBlock != 0) {
            // There is an override, it cannot already be set.
            startBlock = _startBlock;
        } else {
            // There is no override and it is not set in storage. Set it to the block number.
            startBlock = block.number;
        }
    }
```

Reply:
Discussed in slack - I'm a fan of this change. @tynes, does pulling the revert out of the first branch affect the reinitializer? That's the only reason I could think of why this logic was wonky.

Reply:
I do think that this is better, great suggestion @refcell - we will need to update the source anyways to reinit, once all contracts have this storage set then we can remove the concept of the override. There is def some storage layout magic that we need to do